src/core/moves/MetropolisHastingsMove.cpp

```diff
@@ -370,36 +370,50 @@ void MetropolisHastingsMove::performMcmcMove( double prHeat, double lHeat, doubl
     for (auto node: touched_nodes)
         node->touch();
 
+    bool fail_probability = not std::isfinite(ln_hastings_ratio);
+
     double ln_prior_ratio = 0.0;
     double ln_likelihood_ratio = 0.0;
 
+    bool zero_or_nan_to_finite = false;
+
     // compute the probability of the current value for each node
     for (auto node: views::concat(touched_nodes, affected_nodes))
     {
-        if ( RbMath::isAComputableNumber(ln_prior_ratio) &&
-             RbMath::isAComputableNumber(ln_likelihood_ratio) &&
-             RbMath::isAComputableNumber(ln_hastings_ratio) )
-        {
-            double ratio = 0;
-            try {
-                ratio = node->getLnProbabilityRatio();
-            }
-            catch (const RbException &e)
-            {
-                ratio = RbConstants::Double::neginf;
+        if (fail_probability) break;
 
-                if ( e.getExceptionType() != RbException::MATH_ERROR )
-                {
-                    throw;
-                }
-            }
+        if (not node->isStochastic()) continue;
+
+        double ratio = 0;
+        try {
+            // There should be a previous lnProbability because the nodes have been touched.
+            double prev = node->getPrevLnProbability();
+            // Compute the current lnProbability.
+            double current = node->getLnProbability();
 
-            if ( node->isClamped() )
-                ln_likelihood_ratio += ratio;
+            if (std::isfinite(current) and (std::isnan(prev) or prev == RbConstants::Double::neginf))
+            {
+                // If the log-probability changes from NaN or -Inf to something finite,
+                // the ratio would be NaN or +Inf, and the move would be rejected.
+                zero_or_nan_to_finite = true;
+            }
             else
-                ln_prior_ratio += ratio;
+                ratio = current - prev;
         }
+        catch (const RbException &e)
+        {
+            if ( e.getExceptionType() != RbException::MATH_ERROR )
+                throw;
```

Comment:
shouldn't that be throw e ? so we keep the original error

Reply:
It actually does.  This is special syntax to re-throw the exception we just caught.  It is generally safer because it re-uses the existing exception.

Reply:
ok thanks :)